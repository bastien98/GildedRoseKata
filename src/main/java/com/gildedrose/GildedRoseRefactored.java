package com.gildedrose;

import java.util.List;

public class GildedRoseRefactored {
    List<Item> items;

    public GildedRoseRefactored(List<Item> items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            if (item.name.equals("Aged Brie")) {
                if (item.quality < 50) {
                    item.quality = item.quality + 1;
                }
            }

            if(item.quality >0 && !item.name.equals("Sulfuras, Hand of Ragnaros") && !item.name.equals("Backstage passes to a TAFKAL80ETC concert") && !item.name.equals("Aged Brie") ) {
                item.quality = item.quality - 1;
            }

            if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                updateBackstage(item);
            }

            if (!item.name.equals("Sulfuras, Hand of Ragnaros")) {
                item.sellIn = item.sellIn - 1;
            }

            if (item.sellIn < 0) {
                if (item.name.equals("Aged Brie")) {
                    if (item.quality < 50) {
                        item.quality = item.quality + 1;
                    }
                } else {
                    if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                        item.quality = 0;
                    } else {
                        if (item.quality > 0) {
                            if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
                                continue;
                            }
                            item.quality = item.quality - 1;
                        }
                    }
                }
            }
        }
    }

    private void updateBackstage(final Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }

        if (item.sellIn < 11) {
            if (item.quality < 50) {
                item.quality = item.quality + 1;
            }
        }

        if (item.sellIn < 6) {
            if (item.quality < 50) {
                item.quality = item.quality + 1;
            }
        }
    }


}
