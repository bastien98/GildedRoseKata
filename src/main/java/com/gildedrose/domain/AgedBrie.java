package com.gildedrose.domain;


import static com.gildedrose.domain.ItemConfig.MAXIMUM_ITEM_QUALITY;

public class AgedBrie extends Item implements Updatable {

    public AgedBrie(Item item) {
        super(item.name, item.sellIn, item.quality);
    }

    public void update() {
        this.sellIn--;
        if (this.quality < MAXIMUM_ITEM_QUALITY) {
            this.quality++;
        }
        if (this.quality < MAXIMUM_ITEM_QUALITY && this.sellIn < 0) {
            this.quality++;
        }
    }
}
