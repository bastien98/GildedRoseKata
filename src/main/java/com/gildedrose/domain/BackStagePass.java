package com.gildedrose.domain;


import static com.gildedrose.domain.ItemConfig.MAXIMUM_ITEM_QUALITY;
import static com.gildedrose.domain.ItemConfig.MINIMUM_ITEM_QUALITY;

public class BackStagePass extends Item implements SpecializedItem {
    public BackStagePass(Item item) {
        super(item.name, item.sellIn, item.quality);
    }

    public void updateItem() {
        this.sellIn--;

        if (this.quality < MAXIMUM_ITEM_QUALITY) {
            this.quality++;
        }

        if (this.sellIn < 10 && this.quality < MAXIMUM_ITEM_QUALITY) {
            this.quality++;
        }

        if (this.sellIn < 5 && this.quality < MAXIMUM_ITEM_QUALITY) {
            this.quality++;
        }

        if (this.sellIn < 0) {
            this.quality = MINIMUM_ITEM_QUALITY;
        }
    }
}
