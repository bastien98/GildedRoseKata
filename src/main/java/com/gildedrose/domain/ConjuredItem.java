package com.gildedrose.domain;

import static com.gildedrose.domain.ItemConfig.MINIMUM_ITEM_QUALITY;

public class ConjuredItem extends Item implements Updatable {

    public ConjuredItem(Item item) {
        super(item.name, item.sellIn, item.quality);
    }

    @Override
    public void update() {
        this.sellIn--;
        if (this.quality > MINIMUM_ITEM_QUALITY) {
            this.quality = this.quality - 2;
            if (this.quality < MINIMUM_ITEM_QUALITY) {
                this.quality = 0;
            }
        }

        if (this.sellIn < 0 && this.quality > MINIMUM_ITEM_QUALITY) {
            this.quality = this.quality - 2;
        }
    }
}
