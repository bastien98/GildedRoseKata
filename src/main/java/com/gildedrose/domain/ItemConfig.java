package com.gildedrose.domain;

public class ItemConfig {
    public static final int MINIMUM_ITEM_QUALITY = 0;
    public static final int MAXIMUM_ITEM_QUALITY = 50;
}
