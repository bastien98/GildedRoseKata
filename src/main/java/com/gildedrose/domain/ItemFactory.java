package com.gildedrose.domain;

public class ItemFactory {
    public static SpecializedItem createSpecializedItem(Item item) {

        if (isBackStageItem(item.name)) {
            return new BackStagePass(item);
        }
        if ("Aged Brie".equals(item.name)) {
            return new AgedBrie(item);
        }
        if ("Sulfuras, Hand of Ragnaros".equals(item.name)) {
            return new Sulfuras(item);
        }
        if (isConjuredItem(item.name)) {
            return new ConjuredItem(item);
        }
         return new NormalItem(item);
    }

    private static boolean isBackStageItem(String itemName) {
        return itemName.startsWith("Backstage passes");
    }

    private static boolean isConjuredItem(String itemName) {
        return itemName.startsWith("Conjured");
    }
}
