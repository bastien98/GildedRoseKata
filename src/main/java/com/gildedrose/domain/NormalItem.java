package com.gildedrose.domain;


import static com.gildedrose.domain.ItemConfig.MINIMUM_ITEM_QUALITY;

public class NormalItem extends Item implements Updatable {

    public NormalItem(Item item) {
        super(item.name, item.sellIn, item.quality);
    }

    public void update() {
        this.sellIn--;
        if (this.quality > MINIMUM_ITEM_QUALITY) {
            this.quality--;
        }

        if (this.sellIn < 0 && this.quality > MINIMUM_ITEM_QUALITY) {
            this.quality--;
        }
    }
}
