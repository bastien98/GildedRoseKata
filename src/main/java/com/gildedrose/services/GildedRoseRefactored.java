package com.gildedrose.services;

import com.gildedrose.domain.Item;
import com.gildedrose.domain.ItemFactory;
import com.gildedrose.domain.SpecializedItem;

import java.util.ArrayList;
import java.util.List;

public class GildedRoseRefactored {
    private final List<SpecializedItem> items;

    public GildedRoseRefactored(List<Item> items) {
        this.items = new ArrayList<>();
        for (Item item : items) {
            this.items.add(ItemFactory.createSpecializedItem(item));
        }
    }

    public void updateQuality() {
        for (int i = 0; i < items.size(); i++) {
            items.get(i).updateItem();
        }
    }

    public List<SpecializedItem> getItems() {
        return items;
    }
}
