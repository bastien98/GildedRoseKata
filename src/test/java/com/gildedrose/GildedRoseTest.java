package com.gildedrose;

import com.gildedrose.domain.Item;
import com.gildedrose.services.GildedRose;
import com.gildedrose.services.GildedRoseRefactored;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class GildedRoseTest {

    @Test
    void orignalTest() {
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("foo", app.items[0].name);
    }

    @Test
    void abc() {

        // Given
        String[] nameCombos = {
            "+5 Dexterity Vest",
            "Aged Brie",
            "Elixir of the Mongoose",
            "Sulfuras, Hand of Ragnaros",
            "Backstage passes to a TAFKAL80ETC concert"};

        Integer[] sellInCombos = {-1, 0, 5, 6, 7, 10, 11, 12};

        Integer[] qualityCombos = {-1, 0, 1, 49, 50};

        // When + Then
        CombinationApprovals.verifyAllCombinations(
            this::updateItemQuality,
            nameCombos,
            sellInCombos,
            qualityCombos
        );
    }

    private String updateItemQuality(String name, int sellIn, int quality) {
        List<Item> items = new ArrayList<>();
        items.add(new Item(name, sellIn, quality));
        GildedRoseRefactored gildedRoseRefactored = new GildedRoseRefactored(items);
        gildedRoseRefactored.updateQuality();
        return gildedRoseRefactored.items.get(0).toString();
    }


}
