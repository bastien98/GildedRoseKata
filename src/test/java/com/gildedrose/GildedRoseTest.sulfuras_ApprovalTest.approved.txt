[Sulfuras, Hand of Ragnaros, -1, 80] => Sulfuras, Hand of Ragnaros, -1, 80 
[Sulfuras, Hand of Ragnaros, 0, 80] => Sulfuras, Hand of Ragnaros, 0, 80 
[Sulfuras, Hand of Ragnaros, 1, 80] => Sulfuras, Hand of Ragnaros, 1, 80 
[Sulfuras, Hand of Ragnaros, 4, 80] => Sulfuras, Hand of Ragnaros, 4, 80 
[Sulfuras, Hand of Ragnaros, 5, 80] => Sulfuras, Hand of Ragnaros, 5, 80 
[Sulfuras, Hand of Ragnaros, 6, 80] => Sulfuras, Hand of Ragnaros, 6, 80 
[Sulfuras, Hand of Ragnaros, 7, 80] => Sulfuras, Hand of Ragnaros, 7, 80 
[Sulfuras, Hand of Ragnaros, 9, 80] => Sulfuras, Hand of Ragnaros, 9, 80 
[Sulfuras, Hand of Ragnaros, 10, 80] => Sulfuras, Hand of Ragnaros, 10, 80 
[Sulfuras, Hand of Ragnaros, 11, 80] => Sulfuras, Hand of Ragnaros, 11, 80 
